test/general/direct_iterator.pass.cpp

```diff
@@ -0,0 +1,71 @@
+// -*- C++ -*-
+//===-- direct_iterator.pass.cpp ------------------------------------------===//
+//
+// Copyright (C) Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#include <oneapi/dpl/execution>
+#include <oneapi/dpl/algorithm>
+#include <oneapi/dpl/iterator>
+#include <oneapi/dpl/functional>
+
+#include "support/utils.h"
+
+#include <iostream>
+#include <vector>
+#include <numeric>
+
+#if __cpp_lib_span >= 202002L
+#include <span>
+#endif
+
+int
+main()
+{
+#if TEST_DPCPP_BACKEND_PRESENT
+    using T = int;
+
+    const int n = 1000;
+
+    std::vector<T> v(n);
+    std::iota(v.begin(), v.end(), 0);
+
+    sycl::queue q(sycl::default_selector_v);
+
+    T* p = sycl::malloc_device<T>(n, q);
+
+    q.memcpy(p, v.data(), n * sizeof(T)).wait();
+
+    auto v_ref = std::reduce(v.begin(), v.end(), 0);
+
+    dpl::make_direct_iterator d_first(p);
+    dpl::make_direct_iterator d_last(p + n);
+
+    auto v_dev = dpl::reduce(d_first, d_last, 0);
+
+    EXPECT_EQ(v_ref, v_dev);
+
+#if __cpp_lib_span >= 202002L
+
+    std::span<T> x(p, n);
+
+    dpl::make_direct_iterator s_first(x.begin());
```

Comment:
Probably we can avoid this "identical" iterator-wrapper, by introducing just specialization for the trait `oneapi::dpl::__internal::is_passed_directly<_Iter>` ?

like `oneapi::dpl::__internal::is_passed_directly<[_Iter](std::span::iterator)>`, with returns `std::true_type` ?

Reply:
You don't always want to pass in `std::span` iterators directly, since they might not be accessible on the device.  Suppose a user wrote the following:

```cpp
std::vector<int> v(...);
std::span s(v);

// Runtime error, since `s.begin()` is a host iterator and cannot
// be used directly on the device.
dpl::reduce(policy, s.begin(), s.end());
```

You can't in general know whether a span is accessible on the device, and this holds for most ranges you might encounter.  There are a lot of iterator types that users might want to pass into oneDPL directly, and I don't think we can automatically most of them.  I will add a better motivating example below.